lib/src/command/upgrade.dart

```diff
@@ -35,23 +47,241 @@ class UpgradeCommand extends PubCommand {
     argParser.addFlag('precompile',
         help: 'Precompile executables in immediate dependencies.');
 
+    argParser.addFlag('nullsafety',
+        negatable: false,
+        help: 'Upgrade constraints in pubspec.yaml to null-safety versions');
+    argParser.addFlag('null-safety', negatable: false, hide: true);
+
     argParser.addFlag('packages-dir', hide: true);
   }
 
+  /// Avoid showing spinning progress messages when not in a terminal.
+  bool get _shouldShowSpinner => stdout.hasTerminal;
+
+  bool get _dryRun => argResults['dry-run'];
+
   @override
   Future<void> runProtected() async {
     if (argResults.wasParsed('packages-dir')) {
       log.warning(log.yellow(
           'The --packages-dir flag is no longer used and does nothing.'));
     }
+
+    if (argResults['nullsafety'] || argResults['null-safety']) {
+      return await _runUpgradeNullSafety();
+    }
+
     await entrypoint.acquireDependencies(SolveType.UPGRADE,
         useLatest: argResults.rest,
-        dryRun: argResults['dry-run'],
+        dryRun: _dryRun,
         precompile: argResults['precompile']);
 
+    _showOfflineWarning();
+  }
+
+  Future<void> _runUpgradeNullSafety() async {
+    final upgradeOnly = [
+      if (argResults.rest.isNotEmpty)
+        ...argResults.rest
```

Comment:
We need a check that all items in this list are indeed current dependencies.

Reply:
That's a good point!